# Abstract

This notebook deals with the preparation of the data so they can be used as input to the models. It includes __missing values imputation__, __feature engineering__, __encoding__ of categorical variables, __undersampling__ the majority class and __oversampling__ the minority class to create a balanced dataset.

# Import and Basic Info

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import itertools
import math
import random

%matplotlib inline
pd.set_option('display.max_columns', None)
pd.options.mode.chained_assignment = None  # To ignore SettingWithCopyWarning warning

# Set a random seed for repeatability
rand_seed = 123
random.seed(rand_seed)
np.random.seed(rand_seed)

### Define for which dataset the code will run

In [2]:
# This variable denotes we want features for the pCTR predictor or for the base_bid predictor.
name_helper = "pCTR___"

In [3]:
train = pd.read_csv("train.csv")
validation = pd.read_csv("validation.csv")
test = pd.read_csv("test.csv")

In [4]:
train.shape, validation.shape, test.shape

((2430981, 25), (303925, 25), (303375, 22))

In [5]:
train.columns

Index(['click', 'weekday', 'hour', 'bidid', 'userid', 'useragent', 'IP',
       'region', 'city', 'adexchange', 'domain', 'url', 'urlid', 'slotid',
       'slotwidth', 'slotheight', 'slotvisibility', 'slotformat', 'slotprice',
       'creative', 'bidprice', 'payprice', 'keypage', 'advertiser', 'usertag'],
      dtype='object')

In [6]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2430981 entries, 0 to 2430980
Data columns (total 25 columns):
click             int64
weekday           int64
hour              int64
bidid             object
userid            object
useragent         object
IP                object
region            int64
city              int64
adexchange        float64
domain            object
url               object
urlid             float64
slotid            object
slotwidth         int64
slotheight        int64
slotvisibility    object
slotformat        object
slotprice         int64
creative          object
bidprice          int64
payprice          int64
keypage           object
advertiser        int64
usertag           object
dtypes: float64(2), int64(11), object(12)
memory usage: 463.7+ MB


In [7]:
train.head(10)

click  weekday  hour                                     bidid  \
0      0        5    22  b7bea80521fdecd95d2d761a38c91c3f09618066   
1      0        1    20  4f51205475678f5a124bc76b2c54163bf8eaa7eb   
2      0        3    13  b604e3fd054a658ab7ced4285ebf2ef54d2bd890   
3      0        6    23  0348beeae93e561584c3b50fc9e7746a33048ad7   
4      0        5     6  268149c1789bce2bc9798ffd97ec431219bafeb3   
5      0        4    17  1be2cf4a47a2a6aee4b0fa64d1b786d3897be4f0   
6      0        3    12  6b4f23dd215a5f11136c2f29214236cfee1f0122   
7      0        3    16  f7c1d8ea177211249456c79e194617ce107bc077   
8      0        5    14  9e97694096fe8692851048a893231a7f7ff87922   
9      0        5     0  d1b7c7fdb88ce3aa78a540bb29848dd77329a574   

                                     userid       useragent             IP  \
0  2e880fb7d690cf7377b2e42e701728e3f3c0e4c1      windows_ie   125.37.175.*   
1  3a1fe01360ff8100e7d006b83b77a3e4c01d928c  windows_chrome    171.36.92.*   
2  801d18a056b6fe6b06a794aef17fb0d6daff2414      windows_ie    59.46.106.*   
3  0d6eaf2259699990e38a1fc5116f112070b9ecdc      windows_ie  114.250.226.*   
4  a239d9bb642460d974ba67f85e63b8d3e214da0e      windows_ie   183.63.192.*   
5  8810fd690f88ddfb797867e3e08e7816f274bd46      windows_ie   119.45.182.*   
6  fbd1bcaf2b33108d2c7f88ac41f0b0194b508a9b      windows_ie    112.82.93.*   
7  9b539ba886146562a3991fd9f94bb9f9bad1f647      windows_ie    118.75.92.*   
8  ddc9c85b6a5726aafaa4c42239e90fb7a44a88d7      windows_ie   61.185.128.*   
9  63b4f9a3638f81af361ab3c4190fd3e1b5e70fea  windows_chrome    117.9.211.*   

   region  city  adexchange                            domain  \
0       2     2         2.0              trqRTvKaXTKfgg24JKTI   
1     238   239         1.0  20fc675468712705dbf5d3eda94126da   
2      40    41         2.0                trqRTJn7O95I1mKYUV   
3       1     1         1.0           5F97t5E0BTK7XhNrUMpENpn   
4     216   233         2.0  13625cb070ffb306b425cd803c4b7ab4   
5     276   277         1.0               trqRTuxoGQdcFNKbuKz   
6      80    89         3.0               3FKElpuEMusyJqKbuKz   
7      15    19         1.0                  trqRTvp8gIc7gspy   
8     333   342         3.0                ersbQv1RdoTy1m58uG   
9       2     2         3.0                ersbQv1RdoTy1m58uG   

                                url  urlid                        slotid  \
0  6447a7dfa30fe6eb410c91860e7c9b45    NaN                    2015392487   
1  3ddf173a94bd23c326683b6373c75dd4    NaN    mm_10982364_973726_8930541   
2  625d1b5916ea925332c7b326c0574cfa    NaN                    1720123646   
3  dedc488b98ca20707bc9a723957e7d1f    NaN   mm_10027070_118039_10308280   
4  4199d1227c511fc776b76594dabff9f8    NaN                    1120200150   
5  bdbc836b2f86c0391bd2f1e1e50739b3    NaN  mm_10024662_3445902_11178345   
6  6957deb8d83340086ef3cb2abd9c7198    NaN              Fashion_F_Width1   
7   4c70d1fc007d5349286f58da6d08fa9    NaN  mm_33208362_3404787_11119183   
8  cbec43b96a01da4a0e06cc78d483a24f    NaN             ALLINONE_F_Width1   
9  1b2f87d6f3b8b473afe7e9fbe597afca    NaN             ALLINONE_F_Width1   

   slotwidth  slotheight slotvisibility slotformat  slotprice  \
0        200         200              2          0          5   
1        300         250     FourthView         Na          0   
2        250         250              2          0          5   
3        160         600              2          1          0   
4        728          90      OtherView         Na        133   
5        336         280              2          1          0   
6       1000          90              0          0         20   
7        160         600              2          1          0   
8       1000          90              0          0         70   
9       1000          90              0          0         70   

                           creative  bidprice  payprice  \
0  a4f763f78ef3eedfe614263b94a8924e       238         5 

## Missing values

In [8]:
def MissingValues(df):
    
    """
    Function to detect null values
    """
    
    obj_cols = list(df.select_dtypes(include=['object']).columns)
    nulls_list = [df[col].str.match("null").sum()*100/df.shape[0] for col in obj_cols]
    nulls_df = pd.DataFrame(index = obj_cols, columns = ["Nulls_Percentage"], data = nulls_list)
    
    return nulls_df

def ImputeMissingValues(df):
    
    """
    As will be seen, the only column that has null values and will be actually used as a feature
    in our models is "adexchange". All other columns with null values (i.e. domain, url, urlid, keypage)
    will be ignored because they have too many values to be encoded (feature space would be enormous)
    """
    
    ### Remember to impute slot_format and slot_visibility!!!
    
    #df.loc[ df["adexchange"] == "null", "adexchange"] = df["adexchange"].dropna().mode()[0]
    df["adexchange"] = df["adexchange"].fillna(df["adexchange"].dropna().mode()[0])
    df["usertag"] = df["usertag"].fillna("")  ## not optimal
    return df


In [9]:
train.isnull().sum()

click                   0
weekday                 0
hour                    0
bidid                   0
userid                  0
useragent               0
IP                      0
region                  0
city                    0
adexchange          49829
domain             137135
url                 86812
urlid             2430981
slotid                  0
slotwidth               0
slotheight              0
slotvisibility          0
slotformat              0
slotprice               0
creative                0
bidprice                0
payprice                0
keypage            504990
advertiser              0
usertag            497479
dtype: int64

In [10]:
validation.isnull().sum()

click                  0
weekday                0
hour                   0
bidid                  0
userid                 0
useragent              0
IP                     0
region                 0
city                   0
adexchange          6176
domain             16932
url                10850
urlid             303925
slotid                 0
slotwidth              0
slotheight             0
slotvisibility         0
slotformat             0
slotprice              0
creative               0
bidprice               0
payprice               0
keypage            62944
advertiser             0
usertag            62186
dtype: int64

In [11]:
train_2 = ImputeMissingValues(train)
validation_2 = ImputeMissingValues(validation)

In [12]:
train_2.shape, validation_2.shape

((2430981, 25), (303925, 25))

In [13]:
train_2.isnull().sum()

click                   0
weekday                 0
hour                    0
bidid                   0
userid                  0
useragent               0
IP                      0
region                  0
city                    0
adexchange              0
domain             137135
url                 86812
urlid             2430981
slotid                  0
slotwidth               0
slotheight              0
slotvisibility          0
slotformat              0
slotprice               0
creative                0
bidprice                0
payprice                0
keypage            504990
advertiser              0
usertag                 0
dtype: int64

In [14]:
validation_2.isnull().sum()

click                  0
weekday                0
hour                   0
bidid                  0
userid                 0
useragent              0
IP                     0
region                 0
city                   0
adexchange             0
domain             16932
url                10850
urlid             303925
slotid                 0
slotwidth              0
slotheight             0
slotvisibility         0
slotformat             0
slotprice              0
creative               0
bidprice               0
payprice               0
keypage            62944
advertiser             0
usertag                0
dtype: int64

## Feature Engineering

### Usertags

Because estimating the usertags is time consuming, we estimate it once before we proceed for further fearure engineering (so we don't need to recalculate usertags every time we run the Feature Engineering function)

In [15]:
def UsertagCategories(df):
    
    # Drop nan
    df = df["usertag"].dropna().reset_index(drop = True)
    
    # Find unique usertags
    usertags_list = [df[i].split(",") for i in range(df.shape[0])]
    
    # itertools.chain.from_iterable joins a list of lists into a single list
    usertags = np.unique(list(itertools.chain.from_iterable(usertags_list)))
    
    # Remove the empty string ""
    usertags = [tag for tag in usertags if len(tag) > 0]
    
    return usertags

In [16]:
usertags_train = UsertagCategories(train)
usertags_validation = UsertagCategories(validation)

In [17]:
# Check if train and validation/test have the same usertags
list(set(usertags_train) - set(usertags_validation)), list(set(usertags_validation) - set(usertags_train))

([], [])

In [18]:
def FeatureEngineering(df):
    
        # Convert numerical to categorical
        df["weekday_cat"] = df["weekday"].map(lambda x: str(x))
        df["hour_cat"] = df["hour"].map(lambda x: str(x))
        df["region_cat"] = df["region"].map(lambda x: str(x))
        df["city_cat"] = df["city"].map(lambda x: str(x))
        df["adexchange_cat"] = df["adexchange"].map(lambda x: str(x))
        df["advertiser_cat"] = df["advertiser"].map(lambda x: str(x))
        
        # Operating system
        df["os"] = df["useragent"].map(lambda x: x.split("_")[0])
        
        # Browser
        df["browser"] = df["useragent"].map(lambda x: x.split("_")[1])
        
        # Slotarea
        df["slotarea"] = df["slotwidth"]*df["slotheight"]   ####.astype("category")
        
        # Part of the day
        df["part_of_the_day"] = ""
        
        df.loc[(df["weekday"] == 0) & (df["hour"] > 0) & (df["hour"] <=8), "part_of_the_day"] = "Sunday_Night"
        df.loc[(df["weekday"] == 0) & (df["hour"] > 8) & (df["hour"] < 17), "part_of_the_day"] = "Sunday_Morning"
        df.loc[(df["weekday"] == 0) & (((df["hour"] >= 17) & (df["hour"] <= 23)) | ((df["hour"] == 0))), "part_of_the_day"] = "Sunday_Evening"
        
        df.loc[(df["weekday"] == 1) & (df["hour"] > 0) & (df["hour"] <=8), "part_of_the_day"] = "Monday_Night"
        df.loc[(df["weekday"] == 1) & (df["hour"] > 8) & (df["hour"] < 17), "part_of_the_day"] = "Monday_Morning"
        df.loc[(df["weekday"] == 1) & (((df["hour"] >= 17) & (df["hour"] <= 23)) | ((df["hour"] == 0))), "part_of_the_day"] = "Monday_Evening"

        
        df.loc[(df["weekday"] == 2) & (df["hour"] > 0) & (df["hour"] <=8), "part_of_the_day"] = "Tuesday_Night"
        df.loc[(df["weekday"] == 2) & (df["hour"] > 8) & (df["hour"] < 17), "part_of_the_day"] = "Tuesday_Morning"
        df.loc[(df["weekday"] == 2) & (((df["hour"] >= 17) & (df["hour"] <= 23)) | ((df["hour"] == 0))), "part_of_the_day"] = "Tuesday_Evening"

        df.loc[(df["weekday"] == 3) & (df["hour"] > 0) & (df["hour"] <=8), "part_of_the_day"] = "Wednesday_Night"
        df.loc[(df["weekday"] == 3) & (df["hour"] > 8) & (df["hour"] < 17), "part_of_the_day"] = "Wednesday_Morning"
        df.loc[(df["weekday"] == 3) & (((df["hour"] >= 17) & (df["hour"] <= 23)) | ((df["hour"] == 0))), "part_of_the_day"] = "Wednesday_Evening"

        df.loc[(df["weekday"] == 4) & (df["hour"] > 0) & (df["hour"] <=8), "part_of_the_day"] = "Tuesday_Night"
        df.loc[(df["weekday"] == 4) & (df["hour"] > 8) & (df["hour"] < 17), "part_of_the_day"] = "Tuesday_Morning"
        df.loc[(df["weekday"] == 4) & (((df["hour"] >= 17) & (df["hour"] <= 23)) | ((df["hour"] == 0))), "part_of_the_day"] = "Tuesday_Evening"

        df.loc[(df["weekday"] == 5) & (df["hour"] > 0) & (df["hour"] <=8), "part_of_the_day"] = "Friday_Night"
        df.loc[(df["weekday"] == 5) & (df["hour"] > 8) & (df["hour"] < 17), "part_of_the_day"] = "Friday_Morning"
        df.loc[(df["weekday"] == 5) & (((df["hour"] >= 17) & (df["hour"] <= 23)) | ((df["hour"] == 0))), "part_of_the_day"] = "Friday_Evening"

        
        df.loc[(df["weekday"] == 6) & (df["hour"] > 0) & (df["hour"] <=8), "part_of_the_day"] = "Saturday_Night"
        df.loc[(df["weekday"] == 6) & (df["hour"] > 8) & (df["hour"] < 17), "part_of_the_day"] = "Saturday_Morning"
        df.loc[(df["weekday"] == 6) & (((df["hour"] >= 17) & (df["hour"] <= 23)) | ((df["hour"] == 0))), "part_of_the_day"] = "Saturday_Evening"
        
        # Slotprice binning
        df["slotprice_cat"] = 0
        
        df.loc[ df["slotprice"] <= 10, "slotprice_cat"] = 0
        df.loc[ (df["slotprice"] > 10) & (df["slotprice"] <= 50), "slotprice_cat"] = 1
        df.loc[ (df["slotprice"] > 50) & (df["slotprice"] <= 100), "slotprice_cat"] = 2
        df.loc[ df["slotprice"] > 100, "slotprice_cat"] = 3

        
        # Usertags
        for tag in usertags_train:
            col_name = "usertag_" + tag
            df[col_name] = df["usertag"].map(lambda x: 1 if tag in x.split(",") else 0)
        
  
        return df


def DropColumns(df):
    
    columns = ["weekday", "hour", "bidid", "userid", "useragent", "IP", "domain", "url", "urlid", "slotid",
               "slotwidth", "slotheight", "slotprice", "keypage", "usertag", "region", "city", "adexchange", "advertiser"]
    df.drop(columns, axis = 1, inplace = True)
    
    return df


def GetDummies(df):
    
    df = pd.get_dummies(df)
    return df

### Combine train and validation sets

Before applying the feature engineering function and encode categorical variables we will concatenate tarin and validation sets. The reason to do that, is that if we encode train and validation categorical columns separately, there is the possibility that one category that exists in train data, does not exist in validation data. This situation would cause our code to crash, since train and validation sets would consist of different features. 

In [19]:
all_data = pd.concat((train_2, validation_2), axis = 0)

In [20]:
train_2.shape, validation_2.shape, all_data.shape

((2430981, 25), (303925, 25), (2734906, 25))

In [21]:
% time all_data = FeatureEngineering(all_data)

CPU times: user 2min 34s, sys: 5.99 s, total: 2min 40s
Wall time: 2min 40s


In [22]:
all_data.shape

(2734906, 104)

In [23]:
# Drop useless features
all_data = DropColumns(all_data)

In [24]:
all_data.shape

(2734906, 85)

In [25]:
all_data.columns

Index(['click', 'slotvisibility', 'slotformat', 'creative', 'bidprice',
       'payprice', 'weekday_cat', 'hour_cat', 'region_cat', 'city_cat',
       'adexchange_cat', 'advertiser_cat', 'os', 'browser', 'slotarea',
       'part_of_the_day', 'slotprice_cat', 'usertag_10006', 'usertag_10024',
       'usertag_10031', 'usertag_10048', 'usertag_10052', 'usertag_10057',
       'usertag_10059', 'usertag_10063', 'usertag_10067', 'usertag_10074',
       'usertag_10075', 'usertag_10076', 'usertag_10077', 'usertag_10079',
       'usertag_10083', 'usertag_10093', 'usertag_10102', 'usertag_10110',
       'usertag_10111', 'usertag_10114', 'usertag_10115', 'usertag_10116',
       'usertag_10117', 'usertag_10118', 'usertag_10120', 'usertag_10123',
       'usertag_10125', 'usertag_10126', 'usertag_10127', 'usertag_10129',
       'usertag_10130', 'usertag_10131', 'usertag_10133', 'usertag_10138',
       'usertag_10140', 'usertag_10142', 'usertag_10145', 'usertag_10146',
       'usertag_10147', 'usertag

In [26]:
# Convert categorical features to numerical using dummy variables
all_data_dummy = GetDummies(all_data)

In [27]:
c = all_data_dummy.columns
c

Index(['click', 'bidprice', 'payprice', 'slotarea', 'slotprice_cat',
       'usertag_10006', 'usertag_10024', 'usertag_10031', 'usertag_10048',
       'usertag_10052',
       ...
       'part_of_the_day_Saturday_Night', 'part_of_the_day_Sunday_Evening',
       'part_of_the_day_Sunday_Morning', 'part_of_the_day_Sunday_Night',
       'part_of_the_day_Tuesday_Evening', 'part_of_the_day_Tuesday_Morning',
       'part_of_the_day_Tuesday_Night', 'part_of_the_day_Wednesday_Evening',
       'part_of_the_day_Wednesday_Morning', 'part_of_the_day_Wednesday_Night'],
      dtype='object', length=701)

In [28]:
all_data_dummy.head()

click  bidprice  payprice  slotarea  slotprice_cat  usertag_10006  \
0      0       238         5     40000              0              0   
1      0       294        23     75000              0              0   
2      0       238        24     62500              0              1   
3      0       300        25     96000              0              0   
4      0       277       133     65520              3              0   

   usertag_10024  usertag_10031  usertag_10048  usertag_10052  usertag_10057  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              1              0   
3              0              0              0              0              0   
4              0              0              0              0              0   

   usertag_10059  usertag_10063  usertag_10067  usertag_10074  usertag_10075  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              0              0   
3              0              1              0              0              0   
4              0              0              0              0              0   

   usertag_10076  usertag_10077  usertag_10079  usertag_10083  usertag_10093  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              0              0   
3              0              0              0              0              0   
4              0              0              0              0              0   

   usertag_10102  usertag_10110  usertag_10111  usertag_10114  usertag_10115  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              1              0              0              0   
3              0              0              1              0              0   
4              0              0              0              0              0   

   usertag_10116  usertag_10117  usertag_10118  usertag_10120  usertag_10123  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              0              0   
3              0              0              0              0              0   
4              0              0              0              0              0   

   usertag_10125  usertag_10126  usertag_10127  usertag_10129  usertag_10130  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              0              0   
3              0              0              0              0              0   
4              0              0              0              0              0   

   usertag_10131  usertag_10133  usertag_10138  usertag_10140  usertag_10142  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              0              0   
3              0              0              0              0              0   
4              0              0              0              0              0   

   usertag_10145  usertag_10146  usertag_10147  usertag_10148  usertag_10149  \
0              0              0              0              0              0   
1              0              0            

### Split back to train and validation

In [29]:
train_dummy = all_data_dummy[:train_2.shape[0]]
validation_dummy = all_data_dummy[train_2.shape[0]:]

In [30]:
train_dummy.shape, validation_dummy.shape

((2430981, 701), (303925, 701))

In [31]:
train_dummy.head()

click  bidprice  payprice  slotarea  slotprice_cat  usertag_10006  \
0      0       238         5     40000              0              0   
1      0       294        23     75000              0              0   
2      0       238        24     62500              0              1   
3      0       300        25     96000              0              0   
4      0       277       133     65520              3              0   

   usertag_10024  usertag_10031  usertag_10048  usertag_10052  usertag_10057  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              1              0   
3              0              0              0              0              0   
4              0              0              0              0              0   

   usertag_10059  usertag_10063  usertag_10067  usertag_10074  usertag_10075  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              0              0   
3              0              1              0              0              0   
4              0              0              0              0              0   

   usertag_10076  usertag_10077  usertag_10079  usertag_10083  usertag_10093  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              0              0   
3              0              0              0              0              0   
4              0              0              0              0              0   

   usertag_10102  usertag_10110  usertag_10111  usertag_10114  usertag_10115  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              1              0              0              0   
3              0              0              1              0              0   
4              0              0              0              0              0   

   usertag_10116  usertag_10117  usertag_10118  usertag_10120  usertag_10123  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              0              0   
3              0              0              0              0              0   
4              0              0              0              0              0   

   usertag_10125  usertag_10126  usertag_10127  usertag_10129  usertag_10130  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              0              0   
3              0              0              0              0              0   
4              0              0              0              0              0   

   usertag_10131  usertag_10133  usertag_10138  usertag_10140  usertag_10142  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              0              0   
3              0              0              0              0              0   
4              0              0              0              0              0   

   usertag_10145  usertag_10146  usertag_10147  usertag_10148  usertag_10149  \
0              0              0              0              0              0   
1              0              0            

In [32]:
train_dummy.columns

Index(['click', 'bidprice', 'payprice', 'slotarea', 'slotprice_cat',
       'usertag_10006', 'usertag_10024', 'usertag_10031', 'usertag_10048',
       'usertag_10052',
       ...
       'part_of_the_day_Saturday_Night', 'part_of_the_day_Sunday_Evening',
       'part_of_the_day_Sunday_Morning', 'part_of_the_day_Sunday_Night',
       'part_of_the_day_Tuesday_Evening', 'part_of_the_day_Tuesday_Morning',
       'part_of_the_day_Tuesday_Night', 'part_of_the_day_Wednesday_Evening',
       'part_of_the_day_Wednesday_Morning', 'part_of_the_day_Wednesday_Night'],
      dtype='object', length=701)

## Target variables

In [33]:
train["click"].value_counts()

0    2429188
1       1793
Name: click, dtype: int64

In [34]:
validation["click"].value_counts()

0    303723
1       202
Name: click, dtype: int64

## Downsample Training Set

It is obvious that we have to deal with extremely imbalanced datasets. In the training set, the "negative" class (e.g. impressions) corresponds to 99.92% of total data-points and "positive" class corresponds to 0.08%. In absolute values, we have 2429188 impressions and only 1793 clicks. That means that we have a class-imbalance ratio in the order of 1:1000. For more information on the class imbalance problem, check [here](http://www.ele.uri.edu/faculty/he/PDFfiles/ImbalancedLearning.pdf). 

In [35]:
from imblearn.under_sampling import TomekLinks, RandomUnderSampler, NearMiss
from imblearn.over_sampling import SMOTE
from imblearn.combine import SMOTETomek

In [36]:
# Random downsampling of the dataset
if name_helper == "pCTR":
    train_dummy_outliers = train_dummy.loc[train_dummy["click"] == 1]
    train_dummy_inliers = train_dummy.loc[train_dummy["click"] == 0].sample(n = 20000, random_state = rand_seed).reset_index(drop = True)
else:
    train_dummy_outliers = train_dummy.loc[train_dummy["click"] == 1]
    train_dummy_inliers = train_dummy.loc[train_dummy["click"] == 0].sample(n = 40000, random_state = rand_seed).reset_index(drop = True)

In [37]:
train_dummy_inliers.shape, train_dummy_outliers.shape

((40000, 701), (1793, 701))

In [38]:
train_dummy_2 = pd.concat([train_dummy_inliers, train_dummy_outliers], axis = 0)

In [39]:
train_dummy_2.shape

(41793, 701)

In [40]:
if name_helper == "pCTR":
    X = train_dummy_2.loc[:, ((train_dummy_2.columns != "click") & 
                          (train_dummy_2.columns != "payprice") &
                          (train_dummy_2.columns != "bidprice"))].values
    y = train_dummy_2["click"].values

    X_validation = validation_dummy.loc[:, ((validation_dummy.columns != "click") & 
                                            (validation_dummy.columns != "payprice") & 
                                            (validation_dummy.columns != "bidprice"))]
    y_validation = validation_dummy["click"]
    
else:
    X = train_dummy_2.loc[:, ((train_dummy_2.columns != "click") & 
                          (train_dummy_2.columns != "payprice") &
                          (train_dummy_2.columns != "bidprice"))].values
    y = train_dummy_2["payprice"].values

    X_validation = validation_dummy.loc[:, ((validation_dummy.columns != "click") & 
                                            (validation_dummy.columns != "payprice") & 
                                            (validation_dummy.columns != "bidprice"))]
    y_validation = validation_dummy["payprice"]

In [41]:
X.shape, y.shape, X_validation.shape, y_validation.shape

((41793, 698), (41793,), (303925, 698), (303925,))

In [42]:
clicks = train.loc[train["click"] == 1].shape[0]
impressions = train.loc[train["click"] == 0].shape[0]

In [43]:
# SMOTE oversampling
if name_helper == "pCTR":
    sm = SMOTE(ratio = dict({0: train_dummy_inliers.shape[0], 1: train_dummy_inliers.shape[0]}), random_state = rand_seed)
    % time X_resampled, y_resampled = sm.fit_sample(X, y)
else:
    X_resampled, y_resampled = X, y

In [44]:
X_resampled.shape, y_resampled.shape

((41793, 698), (41793,))

### Store results

In [45]:
from sklearn.externals import joblib

In [46]:
if name_helper == "pCTR":
    joblib.dump(X_resampled, "X_train_resampled_pCTR.pkl")
    joblib.dump(y_resampled, "y_train_resampled_pCTR.pkl")
    joblib.dump(X_validation, "X_validation_pCTR.pkl")
    joblib.dump(y_validation, "y_validation_pCTR.pkl")
    
else:
    joblib.dump(X_resampled, "X_train_resampled_basebid.pkl")
    joblib.dump(y_resampled, "y_train_resampled_basebid.pkl")
    joblib.dump(X_validation, "X_validation_basebid.pkl")
    joblib.dump(y_validation, "y_validation_basebid.pkl")